### ISSUE TITLE
Improve Angular Navigation UI Responsiveness and INP Scores by Breaking Up Route Changes

### ISSUE THREAD

**Issue by matthew-whitfield-uk** [angular/angular#51131]
### Which @angular/* package(s) are relevant/related to the feature request?

core, router

### Description

I have found Angular navigation events can often be slow to update the UI when navigating to complex components.
 
It would be nice to be able to break up routing so that a loading bar or similar could be presented at the start of navigation before the new component or module code is loaded and the UI is updated with the new component.
 
When navigating to a lazy loaded module for the first time, this is possible as you can subscribe to the NavigationStart router event and update the UI with a loading bar or spinner to show that the route navigation has started. However, on subsequent navigations to the module, the route change is performed synchronously, so from my knowledge you can’t immediately update the UI – you must wait until the navigation completes which can be slow.
 
This often means Angular apps can feel slow or unresponsive when navigating to complex components or modules, and it also leads to poor Interaction to Next Paint (INP) scores, which is an incoming Core Web Vital metric and will become a Google ranking factor.
 
Here is a minimal repo demonstrating the issue with a 2s timer in place of a complex component. You can use the Chrome Web Vitals extension to view the INP scores drop after the modules are lazy loaded (as the loading bar doesn’t display).
 
[https://github.com/matthew-whitfield-uk/module-routing-inp](https://github.com/matthew-whitfield-uk/module-routing-inp)
 
 
**What is INP?** 
 
Interaction to Next Paint (INP) is a Core Web Vital metric that will be introduced in March 2024. It measures a page’s responsiveness to user interactions and is usually the latency of the longest click, tap, or keyboard interaction. 
 
A Chrome developer’s blog highlighted Angular as the worst-performing framework, with only 19% of Angular apps passing the maximum 200ms INP score on mobile devices:
 
[https://developer.chrome.com/blog/inp-in-frameworks/
](https://developer.chrome.com/blog/inp-in-frameworks/
) 
Even fairly lightweight Angular projects such as Angular.io are currently failing INP on mobile:
 
[https://pagespeed.web.dev/analysis/https-angular-io/1mdibpcdkh?form_factor=mobile ](https://pagespeed.web.dev/analysis/https-angular-io/1mdibpcdkh?form_factor=mobile )

### Proposed solution

Some SPAs present a loading bar on all route navigations (e.g. YouTube desktop / Github) to give immediate user feedback that the navigation has started.
 
Angular could add an option for similar functionality by breaking up the route change process into - Navigation Start,  Update UI with spinner, and then load new component code. This already happens when navigating to a lazy loaded module for the first time (presumably because the HTTP request for the module js is performed asynchronously which allows for a UI update) but subsequent navigations are performed synchronously.  

Subsequent navigations could be moved into a setTimeout to allow the UI to update and give some immediate user feedback.

### Alternatives considered

Using custom route animations doesn't seem to fix this issue, and neither does adding progress bars within modules/components.

**Comment by atscott** [comment]
Essentially a duplicate of https://github.com/angular/angular/issues/42489

**Comment by atscott** [comment]
Quick note here that this is not at all limited to the Angular router. Here are some demos of the issue:

* [Angular](https://stackblitz.com/edit/stackblitz-starters-wsnxsv?file=src%2Fmain.ts)
* [React](https://stackblitz.com/edit/react-elh18j?file=src%2FPages%2FAbout.js)
* [Vue](https://play.vuejs.org/#eNqFVdtu20YQ/ZUpU0AUKpGyEuSBpRWnQVC36CVIgr6UfaDIkciI3GV3l5IMQf+e2Qsvku0EhixyrmfmnF2dvDotWfBFepFX1g0XCk6QCUwVvm0aOMNG8Bom+xYnCXMBgrcKhfMEoX0d3Dqv86VNE9jchPVVffpMEwYQtBJ9m27fa94y5U9eUNpk6s08l07YYoV1U1H+SgfGxc3qHquK62Y/xCG9GnNjvuhBKsHZdvWuFQKZQwxNqoooDp0PTif40TiCTVtVH8gJ57MpE9o6MUv3XcGPBuUfJduB4reJFybe6ldOz3DPa4zDwf9sRromS5/2Vr89yovDrmesebms9U+JBwhdnHPHYb8Y2pfd5ZNs/ok1Fw/3pVT0NXNGW3fM8rxns+dTD2had6R2BsdsR7se6CKutwwS4Ew6NiTcwr96lJPlBSbhhGBxKsaIsmhoe55dhpk9XsYOvXXwf1etBLUaz+ufdMHCriJ6aj/+1PS0QOnxTPJMGB7NoDlu0rbqatPWx/vQUpWZKBtF5FwlmLY1qoLnMrJvAFv+mRv4/rQzASgCF1hxiqBpZeF3Y5tzAm4n+j8pliTddaSXq4OyXHXw6JgsrXHdKsUZ3GVVme1Imj2Ga3XawEcyuyB2mBgkqlYfnUF6RBHNkD99j5yAzr+hZNY/PSvGKz77eJ82MvKNXNrT+SSmIiu0Dhwiq4EtjtYuCL1gtkzwf4viIXBpb97AhEB0a6cxfesZci1VAmlHGfonMPlEctf5DFpDVKET07ri2c6/WSwW05/1+6ZlmSqJFeuopavt8KuUFnYLDYoNF3XKMgwYP/g6F+BQlBWC/8gJc5cYg6mnm39XLj25g16qdI2Vu4s+mXkiiEtGi4T9vOY5VoESZU1KstMmHtTpsUK2VQUZlwvSlb21ukqXerJ6mNdpQ1cXZ6QoM3riHDLx+uOSeIMutDnxCqUaGYVhy5rdNiB+wyHi7lXwMliEOZ3qkTVAWc/Xgh+kuSsTz7Cqa99RUJjjXnFeyXnalM+1eBR49zp4HdyEVbkOqXpYshyPpnZHOo2pJJG5KbdXQ2pJEn/i70bzfzlsSr9xh9+NTYkWe6BZgdnuCfsXebSQPwik6fY4Go6UQHq37vef/sIjPfdOYrGtKPobzo8oedVqjDbsl5blBHsUZ9D+Zjgr2fazfH9UyGQ3lAY6HCJD5LtvjD7AfRm8Gm1RqocKZZBJ/SNHv5Yz0lrJbN6aixxFBMvmCAS2zOEFHTBzRmoqV7L5mtONVkdws2iOxt6keU5gewt1SRiVhRWk8BN97K1tsyvc0K/NOLJYjjsP5QcAWZaNAESwoL+lq+CdvwJxADqd)

In each of these demos, the routed components block the next paint of the browser for an extended period of time after the click event. The poor INP is a true reflection of time between interaction and the next paint. Artificially adding a delay before the routed components attempt to render would effectively "game" the metric.

**Comment by laijuthomas** [comment]
My angular app is also facing serious INP issues (https://capbargain.com/). While checking the angular.dev website using the performance tab in mobile view, I can see INP above 200. 

<img width="1499" alt="Image" src="https://github.com/user-attachments/assets/78c51477-79e8-4942-baf6-95b37660abbd" />

Can anybody help me understand the root cause?

**Comment by SkyZeroZx** [comment]
I think it would be possible to consider using Speculation Rules (although it is currently an experimental API)
https://developer.mozilla.org/en-US/docs/Web/API/Speculation_Rules_API
https://developer.chrome.com/docs/web-platform/implementing-speculation-rules

**Comment by hheexx** [comment]
Will this be fixed with Navigation API ?

**Comment by atscott** [comment]
> Will this be fixed with Navigation API ?

No, likely not. This really doesn't have anything to do with the Router, but rather the fact that the synchronization process is synchronous. It happens to be that Router navigations cause large DOM synchronizations to happen.

As others have discovered, all you need to do to "game" this metric is to introduce a macrotask into the navigation somewhere so the browser has a chance to paint. This can be done by enabling the view transitions feature or by adding something like a `canActivate` guard at the root that just resolves to `true` (e.g. `canActivate: [() => new Promise(r => setTimeout(r))]`)

**Comment by hheexx** [comment]
@atscott I thought with navigation API browser can detect soft navigations as navigations and then it would exclude them from INP.
I know soft navigations are in test in WebVitals library. https://developer.chrome.com/blog/new-soft-navigations-origin-trial
I have view transitions enabled but it did not help with INP on mobile.

**Comment by atscott** [comment]
@hheexx I thought you were referring to the [Navigation API](https://developer.chrome.com/docs/web-platform/navigation-api), which is different from the proposed [Soft Navigation API](https://developer.chrome.com/blog/new-soft-navigations-origin-trial). If I understand correctly, the [proposed heuristic](https://developer.chrome.com/blog/new-soft-navigations-origin-trial#why_a_heuristic) would apply automatically to history-based routing as well as routers using the new Navigation API.

**Comment by hheexx** [comment]
@atscott yea, Soft navigation API will use heuristics and I guess one of it's inputs will be Navigation API.
Question is, will soft navigation API solve angular router INP problems?
